blockchain-etl/ethereum-etl-airflow
# Upgrade to Composer 2 / Airflow 2

## What?
- Upgrade to GCP Composer 2
- Upgrade to Airflow 2

## Why?
Composer 1 is being deprecated at end of March 2023

## How? 
- [x] Add airflow import test and workflow, update other workflows etc.
- [x] Update and pin requirements
- [x] Airflow 2 code changes
  - [x] Jinja macros
    - [x] execution_date -> logical_date
  - [x] Imports
    - [x] BashOperator
    - [x] EmailOperator
    - [x] PythonOperator
    - [x] ExternalTaskSensor
    - [x] PostgresHook
  - [x] Hooks and Operators usage
    - [x] PythonOperator delete provide_context
    - [x] GoogleCloudStorageHook -> GCSHook
    - [x] BigQueryOperator -> BigQueryInsertJobOperator
- [x] Fix local storage limits by moving to GCS Fuse for Composer
- [x] Build Composer dev env using update scripts (shared privately)
- [x] Update local Airflow instance (inc. removing auth requirement)
- [x] Update docs for Composer 2 build

## Testing
- Build dev environment
- Encounter previously seen issue where Composer's install of PyPi packages throws unexpected conflicts which can't be replicated in local installs
  - Fix by pinning conflicting dependency `eth-hash`
- Encounter undocumented issue where scheduler fails to be given access to the whole dags folder by GCS-Fuse when there are a large number of files.
  - Fix by increased Scheduler RAM from 2 to 4GB
- Notice slight change to `airflow_monitoring.py` in latest Composer
  - Update the file in repo
- Run `ethereum_export_dag`: all tasks succeed
- Run `ethereum_load_dag`: all tasks succeed except verifying blocks count (due to lack of backfill)
  - In BigQuery I diffed all tables in `crypto_ethereum` except 'balances', over 2022-09-12 -> 2022-09-14. Everything matched.
- Run `ethereum_parse_common_dag`: tasks started running fine but fail (due to lack of internal dev project)
- Encounter well-known issue where workers sometimes fail to write logs to the logs folder in Composer bucket
  - This is not fixable with `log_fetch_timeout_sec` since this config has been [blocked in Composer 2.2.5](https://cloud.google.com/composer/docs/concepts/airflow-configurations#blocked_configurations_in_airflow_225)
  - I speculatively increased worker RAM from 4 to 8GB to see if it helps

- In general, further resource tweaks should be expected in production
